### DISCUSSION TITLE
download_dsyms does not use App Store Connect API Key

### DISCUSSION THREAD

**Discussion by tmxkevingriesbach** [fastlane/fastlane#17485] [Q&A] [answered]
@joshdholtz I did recently update our company CI to use the new API keys. Thanks for the great work there!

We just noticed unfortunately, that we can't download dsyms from ASC anymore. The system did present an authentication request, which seems to indicate, that `download_dsyms` is using the old login method. I could also not find any parameters for `download_dsyms` that would allow me to pipe in the key. Is this something you are still planning to support in the future?

**Comment by StijnCoolen** [comment]
+1 also interested in this.

**Comment by miguel-wooliex** [comment]
+1

**Comment by guyaumetremblay** [comment]
I'm also very interested in this. I'm configuring the CD for a client and I can't use a login/password. As I understand that this will not be implemented tomorrow, is there any workaround please?

**Comment by tmxkevingriesbach** [comment]
I was just trying to create a fork with a fix for this problem, when I noticed why Josh did not implement this yet: Fastlane is using the old iTunes Connect API to get dSYMs and stuff because the new API does not support it. This old API is obviously not compatible with the ASC API key.  
**So there's probably no way around until Apple implements something new.**

**Reply by joshdholtz** [reply]
This is correct! There is no new API for this yet but I have a radar created for it and the App Store Connect API team is aware we want this 😊

**Reply by moh-abed** [reply]
Any updates regarding this?

**Reply by rogerluan** [reply]
@moh-abed nope 😞 As of App Store Connect API 1.2.

You can check the latest APIs here: https://developer.apple.com/documentation/appstoreconnectapi

**Reply by cynicer** [reply]
Does this mean that with February 2021 we're not able to access this old API without 2FA anymore https://developer.apple.com/support/authentication/ ?

**Reply by dinsen** [reply]
Correct 😞

**Reply by stherold** [reply]
I personally fell back to run this locally on my machine since from time to time and pray that it will make it into the App Store Connect API sooon 🤞 Session is invalidated every day and for Bitrise SSH into their machines yielded the same results...

**Reply by timbodeit** [reply]
@joshdholtz Could you share a radar number that we can reference when filing duplicates?

**Reply by lordzsolt** [reply]
~There's `SPACESHIP_SKIP_2FA_UPGRADE` that you can use until Apple gets their shit together and fixes their API.~
~https://github.com/fastlane/fastlane/issues/18098~
~But please post the radar, so it gets fixed before Apple fully drops Username/Password authentication.~

This no longer works.

**Reply by Buju77** [reply]
Couldn't we simply use `Spaceship::ConnectAPI.auth(key_id:, issuer_id:, filepath:, )` to authenticate instead of `Spaceship::ConnectAPI.login()` in:

https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/download_dsyms.rb#L18

`register_devices()` action does it and it seems to work over there:

https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/register_devices.rb#L41-L50

Or am I missing something?

**Reply by dinsen** [reply]
There is no endpoint, so it will not work before Apple create it and we can implement it into Fastlane.

**Reply by stherold** [reply]
We seem to have version 1.4 now - any news?

**Reply by knox** [reply]
Nothing new regarding symbols in 1.4.1 😿

**Reply by dinsen** [reply]
Still no endpoint in 1.5.0 😤

**Reply by joshdholtz** [reply]
Nope nope, not yet but they know our need for it so hopefully in next version 🤞

**Reply by ghost** [reply]
That may take another year. Apple is so slow sometimes.

**Reply by Blackjacx** [reply]
Maybe they release a new version in-between. They do so sometimes. Is there a radar/feedback already for this? Maybe we should build up some more pressure...

**Reply by joshdholtz** [reply]
My guess is it won't take a year but... I don't know for sure 🤷‍♂️

**Comment by tahirmt** [comment]
I've been using `FASTLANE_SESSION` in my CI pipeline. The only issue is that we have to create a new session every 30 days

**Reply by Shadester** [reply]
Any good solutions for this? Setting the FASTLANE_SESSION on a CI seems like it might not work? How have you done it @tahirmt ?

**Reply by tahirmt** [reply]
The way I have it working is that I have my `FASTLANE_USER` and `FASTLANE_SESSION` environment variables defined on my CI machine and it seems to work pretty good. I just generate a new session every time it expires. I also have `FASTLANE_PASSWORD` set but that's not necessary I think.

**Reply by Blackjacx** [reply]
For `FASTLANE_PASSWORD` there is a bug in place I think, since you can set any non-empty stringbut the variable must not be empty or unset.

**Reply by ChristopherDrum** [reply]
@tahirmt You generate a new session every time it expires, you said. Is that a manual process? You seem to suggest that having those environment variables might keep this automated for you, but it is unclear to me.

**Reply by dinsen** [reply]
@ChristopherDrum Yes it is. 
`fastlane spaceauth -u your@email.com` will generate a session you can pass to `FASTLANE_SESSION`.
Or if you can login to your CI system and run the spaceauth command, it will work.

**Reply by tahirmt** [reply]
@ChristopherDrum it’s a manual process for me unfortunately. Have to do it every 30 days. I couldn’t find a better solution. As @Dinsen said I run the `spaceauth -u <email>` command to generate a new session.

**Reply by tahirmt** [reply]
I would love it if I could use an app specific password for `FASTLANE_PASSWORD`.

**Reply by Blackjacx** [reply]
For me this session expires every day (tried already to SSH into the machine from Bitrise - same result) so I can also run the DSYM upload locally on my mac each day. Just cloned my repo a second time and perform this task there 😅 Hopefully Apple adds this soon to their API ;-)

**Reply by mathaeus** [reply]
I have the same problem as @Blackjacx. The `FASTLANE_SESSION` is expiring every day. There seem to be a few issues talking about this problem https://github.com/fastlane/fastlane/issues/16384.

Therefore also going back to manual DSYM handling for now 😏.

Maybe @tahirmt and @Dinsen are using some additional black magic that make the session last longer :D

**Reply by lordzsolt** [reply]
You can do this, until Apple decides to disable it as well :\ 
https://github.com/fastlane/fastlane/discussions/17485#discussioncomment-352562

**Reply by dinsen** [reply]
haha no black magic here 😇 

We are actually not use the session login right now.
I disabled all calls that don't use the API Key, because our phone number that is added to all our accounts with 2fa was "blocked" for a week.
So I can't say if the session also expiring every day for us.

**Reply by tahirmt** [reply]
I just checked and I haven't updated my session in 28 days and it still works as of 8 hours ago.

**Reply by mathaeus** [reply]
@lordzsolt thanks for the link. Yeah I'm aware of this "skip 2FA" thing, but I honestly don't wanna use it, since we already moved to relying on the ASC API key for all other actions.

**Reply by mathaeus** [reply]
Thanks @tahirmt for confirming. Happy it works for you 😅 🤷‍♂️

**Reply by tahirmt** [reply]
My set up is jinxed haha. This month when the session expired I started getting error that the session is in incorrect format and the character `:` is not escaped. I had this error before but it went away without a session update.

**Reply by novkostya** [reply]
`FASTLANE_SESSION` is working fine for us as long as it's used in the same region it was generated.
We have mac minis in London, Moscow and St.Petersburg. The session I generated in 🇷🇺 works well on Moscow and St.Petersburg boxes, but not on London ones.

**Reply by tahirmt** [reply]
I know. That’s why I use SSH into the GitHub VMs to generate the session.

**Comment by Nonnus** [comment]
+1 following

**Comment by ghost** [comment]
+1

**Comment by CodeStage** [comment]
I don't understand: Why is the `upload_to_testflight ` action supporting API keys but not the `download_dsyms` action?

**Reply by wvteijlingen** [reply]
See https://github.com/fastlane/fastlane/discussions/17485#discussioncomment-125619

**Comment by ghost** [comment]
+1 following

**Comment by Fuqua214** [comment]
I have also same problem if have any solution please help.

[skylight paycards](https://www.skylightpaycard.run/)

**Reply by ghost** [reply]
Please stop such messages. This is spam. I subscribed to find out when a solution comes up!

**Comment by joshdholtz** [comment, marked as answer]
✨ It finally happened! Version 2.197.0 has support for App Store Connect API Key in `download_dsyms`

https://github.com/fastlane/fastlane/releases/tag/2.197.0

**Reply by davdroman** [reply]
I never thought I'd see the day. Huge!! 🚀

**Reply by christhegrand** [reply]
Break out the fucking champagne

**Reply by joshdholtz** [reply]
@christhegrand I had two brandy old fashions. Does that work?

**Reply by christhegrand** [reply]
Close enough!

**Reply by yev-kanivets** [reply]
Huge thanks!

**Reply by icedice** [reply]
This is great news! I see the change in Fastlane, but this would have required an actual change to App Store Connect API right? I have been unable to find this documented anywhere, does anyone have a link for it?

**Comment by flub** [comment]
https://developer.apple.com/documentation/appstoreconnectapi/buildbundle/attributes

**Comment by Blackjacx** [comment]
F.I.N.A.L.L.Y 🙌

Thank you Apple - Thank you much more @joshdholtz  for the fast integration 🎉

**Comment by mathaeus** [comment]
Thanks so much for integrating this so quickly!

Locally it works fine if I pass all the values ✅ 
On CI I still have an issue though ❌

**Has anyone successfully used it on a CI environment yet?**

I'd assume the `download_dsyms` action would read the values for `key_id`, `issuer_id` and `private key` from ENV vars automatically - just like it does for other actions - yet I get the following error:

```
[!] No value found for 'key_id'
```

@joshdholtz is this step of reading the ENV vars missing?

p.s. [the docs](https://docs.fastlane.tools/app-store-connect-api/) could be adapted to `Yes` now too :)

**Reply by PorterHoskins** [reply]
I was able to pass the key with api_key but it waits for the dsyms to appear and fails. If I run the command locally while CI is waiting for the dsyms to appear my local fastlane is able to find the dsyms.

**Reply by mathaeus** [reply]
> I was able to pass the key with api_key

@PorterHoskins you mean [like in the example in the PR description](https://github.com/fastlane/fastlane/pull/19508#issue-1033081763)?

If yes, then yeah this is what I also tried successfully (locally). On CI the values should be read automatically from ENV vars. 
Or do you mean you were able to use ENV vars?

This is how my code looks and like it works for other lanes (e.g. `match`):

```
    if running_on_ci
      # on CI the `key_id`, `issuer_id` and `private key` are read from ENV vars automatically
      # See https://github.com/fastlane/fastlane/issues/17340#issuecomment-737587729
      app_store_connect_api_key(in_house: false)
    else
      ...
```

**Reply by PorterHoskins** [reply]
So I got it to work like 

```
download_dsyms(api_key: app_store_connect_api_key(in_house: false))
```

Some of the other actions work like this. Either way the docs should be updated so we know what to expect 👍

**Reply by mathaeus** [reply]
Ah, so you directly pass it. I usually just create it and then let the action read it from the shared values, which [should also just work fine](https://github.com/fastlane/fastlane/pull/19508/files#diff-a56d80210f6c451c621850db85f8dba70ba2853e957bef159c7b604a3fcdfae1R249). So not sure what I'm missing.

**Reply by joshdholtz** [reply]
@PorterHoskins You should also be able to just do something like...

```rb
app_store_connect_api_key
download_dsyms
```

`download_dsyms` will implicitly pick up the session `app_store_connect_api_key` stores in the shared values. `in_house` is also defaulted to `false` already so you can just omit that part 😊

**Reply by joshdholtz** [reply]
> Thanks so much for integrating this so quickly!
> 
> Locally it works fine if I pass all the values ✅ On CI I still have an issue though ❌
> 
> **Has anyone successfully used it on a CI environment yet?**
> 
> I'd assume the `download_dsyms` action would read the values for `key_id`, `issuer_id` and `private key` from ENV vars automatically - just like it does for other actions - yet I get the following error:
> 
> ```
> [!] No value found for 'key_id'
> ```
> 
> @joshdholtz is this step of reading the ENV vars missing?
> 
> p.s. [the docs](https://docs.fastlane.tools/app-store-connect-api/) could be adapted to `Yes` now too :)

@mathaeus What ENV values are you trying to use? It should be picking them up automatically

**Reply by mathaeus** [reply]
@joshdholtz same ENV vars as we've been using since we migrated to ASC API key:

```
APP_STORE_CONNECT_API_KEY_IS_KEY_CONTENT_BASE64
APP_STORE_CONNECT_API_KEY_ISSUER_ID
APP_STORE_CONNECT_API_KEY_KEY
APP_STORE_CONNECT_API_KEY_KEY_ID
```

And basically we do exactly what you suggested:

```
app_store_connect_api_key
download_dsyms
```

**Reply by mathaeus** [reply]
🤦‍♂️ just realised the ENV vars are not auto-injected, since those are shared ENV vars across projects, but that is specific to our CI setup.

So it should be fine, sorry for the false alarm.

**Reply by joshdholtz** [reply]
> @joshdholtz same ENV vars as we've been using since we migrated to ASC API key:
> 
> ```
> APP_STORE_CONNECT_API_KEY_IS_KEY_CONTENT_BASE64
> APP_STORE_CONNECT_API_KEY_ISSUER_ID
> APP_STORE_CONNECT_API_KEY_KEY
> APP_STORE_CONNECT_API_KEY_KEY_ID
> ```
> 
> And basically we do exactly what you suggested:
> 
> ```
> app_store_connect_api_key
> download_dsyms
> ```

@mathaeus Thanks! I'm fairly certain that is what I used the same setup in my tests too but I will try again just to make sure 😊 I'll get back to you either way when I'm able to try that out again because that should work 💪

**Reply by mathaeus** [reply]
@joshdholtz I can now confirm that it works properly with the ENV vars, sorry again for the wrong alarm! happy weekend 😅
